buildkite/resource_team_member.go

```diff
@@ -3,174 +3,208 @@ package buildkite
-	originalRole := d.Get("role").(string)
-	client := m.(*Client)
-
-	var mutation struct {
-		TeamMemberCreate struct {
-			TeamMemberEdge struct {
-				Node TeamMemberNode
-			}
-		} `graphql:"teamMemberCreate(input: {teamID: $teamId, userID: $userId})"`
-	}
+func (tm *teamMemberResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan, state teamMemberResourceModel
 
-	vars := map[string]interface{}{
-		"teamId": graphql.ID(d.Get("team_id").(string)),
-		"userId": graphql.ID(d.Get("user_id").(string)),
-	}
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
 
-	err := client.graphql.Mutate(context.Background(), &mutation, vars)
-	if err != nil {
-		return diag.FromErr(err)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	updateTeamMember(d, &mutation.TeamMemberCreate.TeamMemberEdge.Node)
+	log.Printf("Creating team member into team %s ...", plan.TeamId.ValueString())
+	apiResponse, err := createTeamMember(
+		tm.client.genqlient,
+		plan.TeamId.ValueString(),
+		plan.UserId.ValueString(),
+		TeamMemberRole(*plan.Role.ValueStringPointer()),
+	)
 
-	// theres a bug in teamMemberCreate that always sets role to MEMBER
-	// so if using MAINTAINER, make a separate call to change the role if necessary
-	if originalRole == "MAINTAINER" {
-		d.Set("role", "MAINTAINER")
-		UpdateTeamMember(ctx, d, m)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to create team member",
+			fmt.Sprintf("Unable to create team member: %s", err.Error()),
+		)
+		return
 	}
 
-	return nil
+	// Update state with values from API response/plan
+	state.Id = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Id)
+	state.Uuid = types.StringValue(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Uuid)
+	state.TeamId = plan.TeamId
+	state.UserId = plan.UserId
+	// The role of the user will by default be "MEMBER" if none was entered in the plan
+	state.Role = types.StringValue(string(apiResponse.TeamMemberCreate.TeamMemberEdge.Node.Role))
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
 }
 
-// ReadTeamMember retrieves a Buildkite team
-func ReadTeamMember(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	client := m.(*Client)
+func (tm *teamMemberResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state teamMemberResourceModel
 
-	var query struct {
-		Node struct {
-			TeamMember TeamMemberNode `graphql:"... on TeamMember"`
-		} `graphql:"node(id: $id)"`
-	}
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
 
-	vars := map[string]interface{}{
-		"id": d.Id(),
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	err := client.graphql.Query(context.Background(), &query, vars)
+	log.Printf("Reading team member %s ...", state.Id.ValueString())
+	apiResponse, err := get(tm.client.genqlient, state.Id.ValueString())
```

Comment:
See comment in `graphql` file; consider a clearer function name.

Reply:
Adjusting in latest commit